packages/react-mongo/react-mongo.ts

```diff
@@ -0,0 +1,177 @@
+    refs.params = { name, args }
+  }
+
+  useEffect(() => {
+    const computation = Tracker.nonreactive(() => (
+      Tracker.autorun(() => {
+        const { name, args } = refs.params
+        if (!name) return
+
+        refs.subscription = Meteor.subscribe(name, ...args)
+
+        const isReady = refs.subscription.ready()
+        if (isReady !== refs.isReady) {
+          refs.isReady = isReady
+          if (refs.updateOnReady) {
+            forceUpdate()
+          }
+        }
+      })
+    ))
+
+    return () => {
+      computation.stop()
+      delete refs.subscription
+    }
+  }, [refs.params])
+
+  return {
+    stop () {
+      refs.subscription?.stop()
+    },
+    ready () {
+      // Ready runs synchronously with render, should not create side effects.
+      refs.updateOnReady = true
+      return refs.isReady
+    }
+  }
+}
+
+const useSubscriptionServer = (): Meteor.SubscriptionHandle => ({
+  stop() {},
+  ready() { return true }
+})
+
+export const useSubscription = (name?: string, ...args: any[]) => (
+  Meteor.isServer
+    ? useSubscriptionServer()
+    : useSubscriptionClient(name, ...args)
+)
+
+type useCursorActions<T> =
+  | { type: 'refresh', data: T[] }
+  | { type: 'addedAt', document: T, atIndex: number }
+  | { type: 'changedAt', document: T, atIndex: number }
+  | { type: 'removedAt', atIndex: number }
+  | { type: 'movedTo', fromIndex: number, toIndex: number }
+
+const useCursorReducer = <T>(data: T[], action: useCursorActions<T>):T[] => {
+  switch (action.type) {
+    case 'refresh':
+      return action.data
+    case 'addedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        action.document,
+        ...data.slice(action.atIndex)
+      ]
+    case 'changedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        action.document,
+        ...data.slice(action.atIndex + 1)
+      ]
+    case 'removedAt':
+      return [
+        ...data.slice(0, action.atIndex),
+        ...data.slice(action.atIndex + 1)
+      ]
+    case 'movedTo':
+      const doc = data[action.fromIndex]
+      const copy = [
+        ...data.slice(0, action.fromIndex),
+        ...data.slice(action.fromIndex + 1)
+      ]
+      copy.splice(action.toIndex, 0, doc)
+      return copy
+  }
+}
+
+const useCursorClient = <T = any>(factory: () => Mongo.Cursor<T>, deps: DependencyList = []) => {
+  // used to prevent double fetch on first run
+  let initialData: T[]
+
+  // On later renders due to deps change, this will fetch the cursor data
+  const cursor = useMemo<Mongo.Cursor<T>>(() => {
+    const c = Tracker.nonreactive(factory)
+    // initialData will only be not falsy on first render, and right after deps change
+    initialData = Tracker.nonreactive(() => c.fetch())
+    return c
+  }, deps)
+
+  // On first render only, this will fetch the cursor data
+  const [data, dispatch] = useReducer<Reducer<T[], useCursorActions<T>>, T[]>(useCursorReducer, undefined, () => initialData)
```

Comment:
super minor (cannot test at the moment), but the lazy-init shape seems unneeded here, it could simply be `useReReducer(useCursorReducer, initialData) ?

Reply:
For some reason, Typescript is giving me crap about that. I think you are right, but I couldn't get Typescript to play nicely with it...